### ISSUE TITLE
TypeScript clarification

### ISSUE THREAD

**Issue by jhechtf** [fastify/fastify-autoload#44]
Tl;DR: Can't get the autoPrefix thing to work with TS files & ts-node.

I've tried to use autoPrefix two ways with TypeScript files, and I cannot seem to find a way to it to work as expected. 

# Things I've tried

## First Attempt
I tried to add the options.prefix to the Autoload call, like so 

```
// server.ts
import * as Fastify from 'fastify';
import * as Autload from 'autoload';
import { join } from 'path';

const fastify = Fastify();
fastify.register(Autoload, {
  dir: join(process.cwd(), 'services'),
  includeTypeScript: true,
  options: {
    prefix: '/api'
  }
}

fastify.listen(3000, '0.0.0.0', (err, address)=> {
  if(err) {
    console.error(err);
    process.exit();
  }
  console.log(`Listening on ${address}`);
})
```

```
// services/user.ts
module.exports = (fastify, opts, next)=>{
  fastify.get('/user', (req, reply)=> { reply.send('hi'); });
  next();
}
```

This only adds in the `/user` route, and not `/api/user`

## Second Attempt

Basically the same server.ts setup, minus the options object.

```
// services/user.ts
module.exports = (fastify, opts, next) => {
  fastify.get('/user', (r, reply)=>{reply.send('hi'); })
  next();
}

module.exports.autoPrefix = '/api';
```
In both cases `/api/user` is not found, but `/user` is. 

Any tips on how to fix this? I'd be happy to submit a pull request for the documentation once I have a solution.

**Comment by mcollina** [comment]
I think there might be a bug in the code! Would you like to send a fix?

**Comment by jhechtf** [comment]
I can look into it, but if you know what it is it may just be faster for you to. I hadn't looked too hard into the code, I just figured maybe I was screwing up somewhere

**Comment by jhechtf** [comment]
In fiddling around some more it seems to actually be an issue with how `fastify-plugin` interacts with `fastify-autoload`. I'll look into it some more as it may be worth noting this for any TypeScript users, but if it is the case then it's likely not any fault of autload

**Comment by mcollina** [comment]
`autoPrefix` works in JavaScript here, so it's likely a bug in the typescript integration.

**Comment by jhechtf** [comment]
the issue seems to come down to the `fastify-plugin` module. It modifies imported objects in a way that seems sort of odd let me see if I can hop a screenshot.

**Comment by jhechtf** [comment]
## Working
![image](https://user-images.githubusercontent.com/8152471/62033287-7a860d00-b1a0-11e9-82d0-452e385a2a94.png)

the first line is for a TS file that is wrapping the exported function in `fp()` from fastify-plugin (in the real-use scenario, I need to use the fastify-plugin plugin because I need access to the decorators added on in other files). The second line is a route that _works_ as it is displayed, but _will not_ work if I wrap it in `fp()` as the object has the same sort of decorator information presented. See the screenshot below.

## Not Working

![image](https://user-images.githubusercontent.com/8152471/62033555-17e14100-b1a1-11e9-9b25-49c473ba64ac.png)

This will no longer work -- if compatibility with the `fastify-plugin` module is expected of this, then this is an issue. If it is not expected, then this is an odd bug, but you are able to get past it if you only use the `fastify-plugin` module only for things that need to be available in other plugins, and not for web-direct things (i.e. splitting out things like db instances into files that use that the `fp()` wrapper and using just a raw export for anything that defines a web route).

**Comment by jhechtf** [comment]
@mcollina  not sure if you saw my other comments.

**Comment by mcollina** [comment]
Compatibility with fastify-plugin is expected. It works in Javascript, so it should work on TS as well.

**Comment by jhechtf** [comment]
Alright, I'll look into what can be done to resolve the issue -- might take me a bit since I'm swamped with my 9-5 right now, and I'm going to have to look into fastify-plugin to determine what's going on from that end.

**Comment by jhechtf** [comment]
**Random Update:**

This appears to be a problem not only with TypeScript, but with autoloads interactions with `fastify-plugin` in general. I was trying to build a quick example application that had some plugins and services, and I found that I could not access a route that I had wrapped in fastify-plugin out of habit. Removing the fastify-plugin wrapper makes the expected route work. 

**Doesn't Work**
when checking `localhost:3000/api/users` => "nothing found"
```
// services/users.js
const fp = require('fastify-plugin');
const AsyncFn = require('./asyncFn');

module.exports = fp((fastify, opts, next) => {
    fastify.get('/', async ()=>{
        return await AsyncFn();
    });
});
module.exports.autoPrefix = '/api/users'
```

**Works**
when checking `localhost:3000/api/users` => expected results.
```
// services/users.js

const AsyncFn = require('./asyncFn');

module.exports = (fastify, opts, next) => {
    fastify.get('/', async ()=>{
        return await AsyncFn();
    });
};
module.exports.autoPrefix = '/api/users';
```

**Comment by mcollina** [comment]
That looks like a bug! Would you like to send a PR to fix it?

**Comment by jhechtf** [comment]
I'll try to. Looking into how it all works together to determine if it's something that should be fixed on fastify plugin's side or autoload's side.

**Comment by mcollina** [comment]
That’s exactly the use that fastify-plugin provides. It’s _built_ for that.

Il giorno lun 26 ago 2019 alle 18:53 Jim Burbridge <notifications@github.com>
ha scritto:

**Comment by jhechtf** [comment]
I decided to test my hypothesis by creating a quick example that did not use fastify autoload or plugin but did use the `Symbol.for('skip-override')`.  

The results are the same -- a plugin called with the skip-override set to true ignores it's passed prefix, and a plugin registered without it does not. I'm assuming this bug is therefore an issue with the fastify base itself, and not with `fastify-autoload` so I'll be closing this issue.

**Comment by mcollina** [comment]
@jhechtf it's not a bug, it's how fastify works. It's documented in https://www.fastify.io/docs/latest/Plugins/#route-prefixing-option.

**Comment by mrob11** [comment]
@jhechtf I just started running into this problem today in a new project. Did you ever get around this?

**Comment by diego-betto** [comment]
> @jhechtf it's not a bug, it's how fastify works. It's documented in https://www.fastify.io/docs/latest/Plugins/#route-prefixing-option.

Now we get 404 there
